### PR TITLE
GH-36953: [MATLAB] Add gateway `arrow.array` function to create Arrow Arrays from MATLAB data

### LINKED ISSUE #36953: [MATLAB] Add gateway `arrow.array` function to create Arrow Arrays from MATLAB data
### Describe the enhancement requested

As discussed in #36855, we think it would be better to move the recommended APIs for the MATLAB Interface directly under the top-level `arrow.*` package. This should help simplify the interface, and will make it easier for users to switch between multiple language bindings. We have already moved the `type` convenience constructors to the `arrow` package.  Now we want to add a gateway function that creates arrays to mirror `PyArrow`. As part of this change, we will modify the array constructors to accept `libmexclass.proxy.Proxy` objects - similar to how the `arrow.type.<Type>` constructors accept  `libmexclass.proxy.Proxy` objects.

 

### Component(s)

MATLAB

### PR DESCRIPTION
**Pull request by sgilmore10**
### Rationale for this change

As discussed in #36855, we think it would be better to move the recommended APIs for the MATLAB Interface directly under the top-level `arrow.*` package. This should help simplify the interface, and will make it easier for users to switch between multiple language bindings. We have already moved the `type` convenience constructors to the `arrow` package.  Now we want to add a gateway function that creates arrays to mirror `PyArrow`. As part of this change, we will modify the array constructors to accept `libmexclass.proxy.Proxy` objects - similar to how the `arrow.type.<Type>` constructors accept  `libmexclass.proxy.Proxy` objects.

### What changes are included in this PR?

1. Added `arrow.array()` gateway function that can be used to construct arrays:

```matlab
>> arrowArray = arrow.array([1 2 3 4]);
>> class(arrowArray)

ans =

    'arrow.array.Float64Array'

>> arrowArray = arrow.array(["A" "B" "C"]);
>> class(arrowArray)

ans =

    'arrow.array.StringArray'

```

2. Added a static `fromMATLAB()` method to all  subclasses of`arrow.array.Array`. 

```matlab
>> array = arrow.array.StringArray.fromMATLAB(["A" "B" "C"])

array = 

[
  "A",
  "B",
  "C"
]

>> array = arrow.array.TimestampArray.fromMATLAB(datetime(2023, 8, 1))

array = 

[
  2023-08-01 00:00:00.000000
]

```

As part of this change, users can no longer use the `arrow.array.Array` subclass constructors to create arrays. Instead, they can use either `arrow.array()` or the static `fromMATLAB` method.

### Are these changes tested?

Updated the existing tests to account for the API changes and added the following new test classes:

1. arrow/internal/validate/tType.m
2. arrow/internal/validate/tShape.m
3. arrow/internal/validate/tRealNumeric.m
4. arrow/internal/validate/tNonsparse.m
5. arrow/internal/validate/tNumeric.m
6. arrow/array/tArray.m

### Are there any user-facing changes?

Yes, we changed the signature of all `arrow.array.Array` subclasses to accept scalar `libmexclass.proxy.Proxy` classes. NOTE: The MATLAB interface is still under active development. 

### Future Directions

1. In a followup PR, we plan on adding a new name-value pair to `arrow.array()` called `Type`, which can be set to an `arrow.type.Type` object. This will let users specify what kind of arrow array they would like to create from MATLAB data.

* Closes: #36953

### CODE REVIEW

**Review by kou — APPROVED** [review]
+1

(This is too large to review...)

**Comment by sgilmore10** [comment]
Hi @kou, I apologize for the size of this review. I'll make sure to keep my future PRs as small as possible.

I thought about splitting these changes across multiple PRs, but couldn't find a way to do so without mass test failures. 

Again, I apologize for the size of the PR. If it helps, I've summarized the major changes in this PR below:

- Added a new gateway function called `arrow.array()`. This mirrors the design of PyArrow. 
- Added a static `fromMATLAB` method to all concrete subclasses of `arrow.array.Array`. Clients can use these methods if they prefer not to rely on the gateway function.
- We no longer expect or want clients to call the constructors of `arrow.array.Array` subclasses directly. This also mirrors the design of PyArrow.

Best,
Sarah

**Comment by kou** [comment]
Thanks for the summary.
(I should have written as "I could understand what you did but this is too large to review...". Sorry.)

We should wait for a review from @kevingurney .

File: matlab/src/matlab/+arrow/+array/Array.m
Comment (kevingurney) [inline matlab/src/matlab/+arrow/+array/Array.m:35]: Indent?

File: matlab/src/matlab/+arrow/+array/TimestampArray.m
Comment (kevingurney) [inline matlab/src/matlab/+arrow/+array/TimestampArray.m:79]: Maybe we should call this variable `epochTime`?

File: matlab/src/matlab/+arrow/array.m
Comment (kevingurney) [inline matlab/src/matlab/+arrow/array.m:23]: Maybe we should rename this to `convertCellstrToString`?

File: matlab/src/matlab/+arrow/array.m
Comment (kevingurney) [inline matlab/src/matlab/+arrow/array.m:30]: This is a minor comment, but in a lot of other places we have been ordering this list of types as `uint8`, `uint16`, ... rather than "interleaving" the signed and unsigned integer types.

File: matlab/test/arrow/array/hNumericArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/hNumericArray.m:105]: `ND` -> `NDimensional`?

File: matlab/test/arrow/array/tArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tArray.m:40]: "...with respect to the input data array's MATLAB class type."

File: matlab/test/arrow/array/tArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tArray.m:66]: I think this should say something like "Verify that Valid contains the expected logical vector when InferNulls=true."

File: matlab/test/arrow/array/tArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tArray.m:81]: "Verify the Valid property returns the expected logical vector..."

File: matlab/test/arrow/array/tBooleanArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tBooleanArray.m:142]: ND -> NDimensional?

File: matlab/test/arrow/array/tStringArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tStringArray.m:142]: ND -> NDimensional?

File: matlab/test/arrow/array/tTimestampArray.m
Comment (kevingurney) [inline matlab/test/arrow/array/tTimestampArray.m:176]: ND -> NDimensional?

File: matlab/test/arrow/internal/validate/tRealNumeric.m
Comment (kevingurney) [inline matlab/test/arrow/internal/validate/tRealNumeric.m:20]: I think this can be deleted.

File: matlab/test/arrow/internal/validate/tRealNumeric.m
Comment (kevingurney) [inline matlab/test/arrow/internal/validate/tRealNumeric.m:25]: Could we change this to not be "interleaved"?

File: matlab/test/arrow/internal/validate/tShape.m
Comment (kevingurney) [inline matlab/test/arrow/internal/validate/tShape.m:1]: TREALNUMERIC -> TSHAPE

File: matlab/test/arrow/internal/validate/tShape.m
Comment (kevingurney) [inline matlab/test/arrow/internal/validate/tShape.m:21]: 2D -> 2Dimensional?

**Review by kevingurney — CHANGES_REQUESTED** [review]
Thanks for working on this, @sgilmore10!

Overall, the changes look good to me - just some minor comments.

**Comment by sgilmore10** [comment]
> Thanks for the summary. (I should have written as "I could understand what you did but this is too large to review...". Sorry.)
> 
> We should wait for a review from @kevingurney .

Hi @Kou,

 Looking back, I realize I could have held off on making the changes to the validation functions. That would have reduced the size of this review. Going forward I'll make sure keep my PRs as small as possible. I appreciate the feedback!

Best,
Sarah

File: matlab/src/matlab/+arrow/+array/Array.m
Comment (sgilmore10) [reply matlab/src/matlab/+arrow/+array/Array.m:35]: fixed.

File: matlab/src/matlab/+arrow/array.m
Comment (sgilmore10) [reply matlab/src/matlab/+arrow/array.m:23]: I agree.

File: matlab/src/matlab/+arrow/+array/TimestampArray.m
Comment (sgilmore10) [reply matlab/src/matlab/+arrow/+array/TimestampArray.m:79]: done.

File: matlab/src/matlab/+arrow/array.m
Comment (sgilmore10) [reply matlab/src/matlab/+arrow/array.m:30]: That's a good point. I'll reorder the cases.

File: matlab/test/arrow/array/hNumericArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/hNumericArray.m:105]: done.

File: matlab/test/arrow/array/tArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/tArray.m:81]: dibe,

File: matlab/test/arrow/array/tBooleanArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/tBooleanArray.m:142]: done.

File: matlab/test/arrow/array/tStringArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/tStringArray.m:142]: done.

File: matlab/test/arrow/internal/validate/tShape.m
Comment (sgilmore10) [reply matlab/test/arrow/internal/validate/tShape.m:1]: oops. Thanks for catching this.

File: matlab/test/arrow/internal/validate/tShape.m
Comment (sgilmore10) [reply matlab/test/arrow/internal/validate/tShape.m:21]: done.

File: matlab/test/arrow/array/tArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/tArray.m:40]: done.

File: matlab/test/arrow/array/tBooleanArray.m
Comment (sgilmore10) [reply matlab/test/arrow/array/tBooleanArray.m:21]: good point. done.

**Review by kevingurney — APPROVED** [review]

**Comment by kevingurney** [comment]
+1

**Pull request merged**
